server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java

```diff
@@ -1389,19 +1390,17 @@ private static Number value(XContentParser parser, NumberType numberType, Number
 
     private void indexValue(DocumentParserContext context, Number numericValue) {
         List<Field> fields = fieldType().type.createFields(fieldType().name(), numericValue, indexed, hasDocValues, stored);
-        if (dimension) {
-            // Check that a dimension field is single-valued and not an array
-            if (context.doc().getByKey(fieldType().name()) != null) {
-                throw new IllegalArgumentException("Dimension field [" + fieldType().name() + "] cannot be a multi-valued field.");
-            }
-            if (fields.size() > 0) {
-                // Add the first field by key so that we can validate if it has been added
-                context.doc().addWithKey(fieldType().name(), fields.get(0));
-                context.doc().addAll(fields.subList(1, fields.size()));
-            }
-        } else {
-            context.doc().addAll(fields);
+        if (dimension && fields.isEmpty() == false) {
```

Comment:
I think it might be a bit confusing to check if `fields` is empty but then us `numericValue`. Is there something up with default values here? I wonder if we're better off just checking if `numericValue != null`.